rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/BundleResource.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.brooklyn.util.osgi.VersionedName;
+import org.apache.brooklyn.util.osgi.VersionedName.VersionedNameComparator;
+import org.apache.brooklyn.util.yaml.Yamls;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+
+@HaHotStateRequired
+public class BundleResource extends AbstractBrooklynRestResource implements BundleApi {
+
+    private static final Logger log = LoggerFactory.getLogger(BundleResource.class);
+    private static final String LATEST = "latest";
+
+    @Override
+    public List<BundleSummary> list(String versions) {
+        return list(TypeResource.isLatestOnly(versions, true), Predicates.alwaysTrue());
+    }
+    
+    private List<BundleSummary> list(boolean onlyLatest, Predicate<String> symbolicNameFilter) {
+        
+        Map<VersionedName,ManagedBundle> bundles = new TreeMap<>(VersionedNameComparator.INSTANCE);
+        for (ManagedBundle b: ((ManagementContextInternal)mgmt()).getOsgiManager().get().getManagedBundles().values()) {
+            if (symbolicNameFilter.apply(b.getSymbolicName())) {
+                // TODO entitlements for bundles
+                VersionedName key = onlyLatest ? new VersionedName(b.getSymbolicName(), LATEST) : b.getVersionedName();
+                ManagedBundle oldBundle = bundles.get(key);
+                if (oldBundle==null || oldBundle.getVersionedName().compareTo(b.getVersionedName()) > 0) {
+                    bundles.put(key, b);
+                }
+            }
+        }
+        return toBundleSummary(bundles.values());
+    }
+
+    private List<BundleSummary> toBundleSummary(Iterable<ManagedBundle> sortedItems) {
+        List<BundleSummary> result = MutableList.of();
+        for (ManagedBundle t: sortedItems) {
+            result.add(TypeTransformer.bundleSummary(brooklyn(), t, ui.getBaseUriBuilder(), mgmt()));
+        }
+        return result;
+    }
+
+    @Override
+    public List<BundleSummary> listVersions(String symbolicName) {
+        return list(false, Predicates.equalTo(symbolicName));
+    }
+
+    @Override
+    public BundleSummary detail(String symbolicName, String version) {
+        ManagedBundle b = lookup(symbolicName, version);
+        return TypeTransformer.bundleDetails(brooklyn(), b, ui.getBaseUriBuilder(), mgmt());
+    }
+
+    protected ManagedBundle lookup(String symbolicName, String version) {
+        // TODO entitlements for bundles
+
+        ManagedBundle b = ((ManagementContextInternal)mgmt()).getOsgiManager().get().getManagedBundle(new VersionedName(symbolicName, version));
+        if (b==null) {
+            throw WebResourceUtils.notFound("Bundle with id '%s:%s' not found", symbolicName, version);
+        }
+        return b;
+    }
+
+    @Override
+    public BundleInstallationRestResult remove(String symbolicName, String version, Boolean force) {
+        ManagedBundle b = lookup(symbolicName, version);
```

Comment:
Really want an entitlements check (as per your previous TODOs). Otherwise the presence of the `/bundles` endpoint will be a horrible backdoor for allowing a user to damage other people's apps.

Reply:
agree, as above